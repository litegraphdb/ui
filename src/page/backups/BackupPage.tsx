'use client';
import React, { useState } from 'react';
import { LoadingOutlined, PlusSquareOutlined, RedoOutlined } from '@ant-design/icons';
import PageContainer from '@/components/base/pageContainer/PageContainer';
import LitegraphButton from '@/components/base/button/Button';
import LitegraphTable from '@/components/base/table/Table';
import FallBack from '@/components/base/fallback/FallBack';
import { tableColumns } from './constant';
import DeleteBackup from './components/DeleteBackup';
import AddEditBackup from './components/AddEditBackup';
import { downloadBase64File } from '@/utils/appUtils';
import { toast } from 'react-hot-toast';
import { globalToastId } from '@/constants/config';
import { useReadAllBackupsQuery, useReadBackupMutation } from '@/lib/store/slice/slice';
import { BackupMetaData } from 'litegraphdb/dist/types/types';
import LitegraphFlex from '@/components/base/flex/Flex';
import LitegraphText from '@/components/base/typograpghy/Text';
import LitegraphTooltip from '@/components/base/tooltip/Tooltip';

const BackupPage = () => {
  const [isDeleteBackupVisible, setIsDeleteBackupVisible] = useState(false);
  const [isAddEditBackupVisible, setIsAddEditBackupVisible] = useState(false);
  const [selectedBackup, setSelectedBackup] = useState<BackupMetaData | null>(null);
  const {
    data: backupsList = [],
    refetch: fetchBackupsList,
    isLoading,
    isFetching,
    error,
  } = useReadAllBackupsQuery();
  const [fetchBackupByFilename, { isLoading: isDownloading }] = useReadBackupMutation();
  const isBackupsLoading = isLoading || isFetching;
  const handleCreateBackup = () => {
    setSelectedBackup(null);
    setIsAddEditBackupVisible(true);
  };

  const handleDeleteBackup = (backup: BackupMetaData) => {
    setSelectedBackup(backup);
    setIsDeleteBackupVisible(true);
  };

  const handleDownload = async (backup: BackupMetaData) => {
    if (!backup.Filename) {
      toast.error('Missing backup filename', { id: globalToastId });
      return;
    }
    const { data } = await fetchBackupByFilename(backup.Filename);
    if (data && data.Data) {
      downloadBase64File(data.Data, backup.Filename);
    } else {
      toast.error('Unable to download backup', { id: globalToastId });
    }
  };

  return (
    <PageContainer
      id="backups"
      pageTitle={
        <LitegraphFlex align="center" gap={10}>
          <LitegraphText>Backups</LitegraphText>
          {isBackupsLoading ? (
            <LoadingOutlined className="loading-icon" />
          ) : (
            <LitegraphTooltip title="Refresh Data" placement="right">
              <RedoOutlined className="cursor-pointer" onClick={fetchBackupsList} />
            </LitegraphTooltip>
          )}
        </LitegraphFlex>
      }
      pageTitleRightContent={
        <LitegraphButton
          type="link"
          icon={<PlusSquareOutlined />}
          onClick={handleCreateBackup}
          weight={500}
        >
          Create Backup
        </LitegraphButton>
      }
    >
      {error && !isBackupsLoading ? (
        <FallBack retry={fetchBackupsList}>Something went wrong.</FallBack>
      ) : (
        <LitegraphTable
          loading={isBackupsLoading || isDownloading}
          columns={tableColumns(handleDeleteBackup, handleDownload, isDownloading)}
          dataSource={backupsList}
          rowKey={'Filename'}
        />
      )}

      {isAddEditBackupVisible && (
        <AddEditBackup
          isAddEditBackupVisible={isAddEditBackupVisible}
          setIsAddEditBackupVisible={setIsAddEditBackupVisible}
          backup={selectedBackup || null}
        />
      )}

      {isDeleteBackupVisible && selectedBackup && (
        <DeleteBackup
          title={`Are you sure you want to delete "${selectedBackup.Filename}" backup?`}
          paragraphText={'This action will delete backup.'}
          isDeleteModelVisible={isDeleteBackupVisible}
          setIsDeleteModelVisible={setIsDeleteBackupVisible}
          selectedBackup={selectedBackup}
          setSelectedBackup={setSelectedBackup}
        />
      )}
    </PageContainer>
  );
};

export default BackupPage;
