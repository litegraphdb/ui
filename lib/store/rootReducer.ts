import { UnknownAction, combineReducers } from '@reduxjs/toolkit';
import liteGraphReducer from './litegraph/reducer';
import { rtkQueryErrorLogger } from './rtkApiMiddlewear';
import { localStorageKeys, paths } from '@/constants/constant';
import graphReducer from './graph/reducer';
import nodeReducer from './node/reducer';
import edgeReducer from './edge/reducer';
import { LitegraphAction } from './litegraph/actions';
import tenantsReducer from './tenants/reducers';
import tagReducer from './tag/reducer';
import labelReducer from './label/reducer';
import credentialReducer from './credential/reducer';
import userReducer from './user/reducer';
import { tenantLists } from './tenants/actions';
import vectorReducer from './vector/reducer';
import backupReducer from './backup/reducer';
const rootReducer = combineReducers({
  liteGraph: liteGraphReducer,
  graphsList: graphReducer,
  nodesList: nodeReducer,
  edgesList: edgeReducer,
  tenants: tenantsReducer,
  tagsList: tagReducer,
  vectorsList: vectorReducer,
  labelsList: labelReducer,
  credentialsList: credentialReducer,
  usersList: userReducer,
  tenantsList: tenantsReducer,
  backupsList: backupReducer,
});

export const apiMiddleWares = [rtkQueryErrorLogger];

export const handleLogout = (path?: string) => {
  localStorage.removeItem(localStorageKeys.token);
  localStorage.removeItem(localStorageKeys.tenant);
  localStorage.removeItem(localStorageKeys.user);
  localStorage.removeItem(localStorageKeys.adminAccessKey);
  localStorage.removeItem(localStorageKeys.serverUrl);
  window.location.href = path ? path : paths.login;
};

const resettableRootReducer = (
  state: ReturnType<typeof rootReducer> | undefined,
  action: UnknownAction
) => {
  if (action.type === LitegraphAction.LOG_OUT) {
    handleLogout(action.payload as any);
  }
  return rootReducer(state, action);
};

export default resettableRootReducer;
