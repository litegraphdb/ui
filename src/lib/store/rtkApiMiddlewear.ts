import { isRejectedWithValue, Middleware, MiddlewareAPI } from '@reduxjs/toolkit';
import { isNumber } from 'lodash';
import toast from 'react-hot-toast';
import { handleLogout } from './rootReducer';
import { globalToastId } from '@/constants/config';

export const errorHandler = (er: any) => {
  const error = er?.payload || {};
  const endpointName = er?.meta?.arg?.endpointName;
  const serverErrorMessage = error?.data?.message || error?.data?.detail;
  if (isNumber(error?.status) || isNumber(error?.StatusCode)) {
    switch (error.status || error.StatusCode) {
      case 401:
      case 403:
        if (endpointName !== 'login') {
          toast.error('Invalid session, login again. ', { id: globalToastId });
          setTimeout(() => {
            handleLogout();
          }, 2000);
          break;
        }
      default:
        toast.error(serverErrorMessage ? serverErrorMessage : 'Something went wrong.', {
          id: globalToastId,
        });
    }
  }
  if (error?.data == 'Network Error') {
    toast.error('Network Error', { id: globalToastId });
  }
};

export const rtkQueryErrorLogger: Middleware =
  (_api: MiddlewareAPI) => (next: (action: any) => any) => (action: any) => {
    // RTK Query uses `createAsyncThunk` from redux-toolkit under the hood, so we're able to utilize these matchers!
    if (isRejectedWithValue(action)) {
      errorHandler(action);
    }

    return next(action);
  };
