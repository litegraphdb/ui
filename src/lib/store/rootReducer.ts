import { UnknownAction, combineReducers } from '@reduxjs/toolkit';
import liteGraphReducer from './litegraph/reducer';
import { rtkQueryErrorLogger } from './rtkApiMiddlewear';
import { localStorageKeys, paths } from '@/constants/constant';
import graphReducer from './graph/reducer';
import nodeReducer from './node/reducer';
import edgeReducer from './edge/reducer';
import { LitegraphAction } from './litegraph/actions';
import tenantsReducer from './tenants/reducers';
import tagReducer from './tag/reducer';
import labelReducer from './label/reducer';
import credentialReducer from './credential/reducer';
import userReducer from './user/reducer';
import { tenantLists } from './tenants/actions';
import vectorReducer from './vector/reducer';
import backupReducer from './backup/reducer';
import sdkSlice from './rtk/rtkSdkInstance';
const rootReducer = combineReducers({
  [sdkSlice.reducerPath]: sdkSlice.reducer,
  liteGraph: liteGraphReducer,
  graphsList: graphReducer,
  nodesList: nodeReducer,
  edgesList: edgeReducer,
  tenants: tenantsReducer,
  tagsList: tagReducer,
  vectorsList: vectorReducer,
  labelsList: labelReducer,
  credentialsList: credentialReducer,
  usersList: userReducer,
  tenantsList: tenantsReducer,
  backupsList: backupReducer,
});

export const apiMiddleWares = [rtkQueryErrorLogger, sdkSlice.middleware];

export const handleLogout = (path?: string) => {
  localStorage.removeItem(localStorageKeys.token);
  localStorage.removeItem(localStorageKeys.tenant);
  localStorage.removeItem(localStorageKeys.user);
  localStorage.removeItem(localStorageKeys.adminAccessKey);
  localStorage.removeItem(localStorageKeys.serverUrl);
  window.location.href = path ? path : paths.login;
};

const resettableRootReducer = (
  state: ReturnType<typeof rootReducer> | undefined,
  action: UnknownAction
) => {
  if (action.type === LitegraphAction.LOG_OUT) {
    handleLogout(action.payload as any);
  }
  return rootReducer(state, action);
};

export default resettableRootReducer;
