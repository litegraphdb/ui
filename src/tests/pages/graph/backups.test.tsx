import '@testing-library/jest-dom';
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMockStore } from '../../store/mockStore';
import { mockBackupData } from '../mockData';
import BackupPage from '@/page/backups/BackupPage';

jest.mock('antd', () => jest.requireActual('antd'));

jest.mock('@/hooks/entityHooks', () => ({
  useBackups: () => ({
    backupsList: mockBackupData,
    isLoading: false,
    error: null,
    fetchBackupsList: jest.fn(),
  }),
}));

describe('BackupPage', () => {
  const store = createMockStore();

  it('should render the BackupPage', () => {
    render(
      <Provider store={store}>
        <BackupPage />
      </Provider>
    );

    const titleElement = screen.getByText('Backups');
    expect(titleElement).toBeInTheDocument();
    expect(titleElement).toMatchSnapshot();
  });

  it('should display Create Backup button', () => {
    render(
      <Provider store={store}>
        <BackupPage />
      </Provider>
    );

    const createButton = screen.getByRole('button', { name: /create backup/i });
    expect(createButton).toBeInTheDocument();
    expect(createButton).toBeVisible();
    expect(createButton).toMatchSnapshot();
  });

  it('should create a backup and should be visible in the table', async () => {
    jest.setTimeout(15000);

    const { container } = render(
      <Provider store={store}>
        <BackupPage />
      </Provider>
    );

    const initialTable = container.querySelector('.ant-table');
    expect(initialTable).toMatchSnapshot('initial table state');

    const createButton = screen.getByRole('button', { name: /create backup/i });
    expect(createButton).toMatchSnapshot('create backup button');
    fireEvent.click(createButton);

    const addBackupModal = container.querySelector('.ant-modal');
    expect(addBackupModal).toMatchSnapshot('add backup modal');

    const backupNameInput = screen.getByLabelText('Filename');
    expect(backupNameInput).toMatchSnapshot('backup name input');

    const backupName = mockBackupData[0].Filename;
    fireEvent.change(backupNameInput, { target: { value: backupName } });

    const submitButton = screen.getByRole('button', { name: /ok/i });
    expect(submitButton).toMatchSnapshot('submit button');
    fireEvent.click(submitButton);

    const finalTable = container.querySelector('.ant-table');
    expect(finalTable).toMatchSnapshot('final table state');
  }, 15000);

  it('should delete a backup and should be removed from the table', async () => {
    jest.setTimeout(15000);

    const { container } = render(
      <Provider store={store}>
        <BackupPage />
      </Provider>
    );

    const initialTable = container.querySelector('.ant-table');
    expect(initialTable).toMatchSnapshot('initial table state');

    const menuButtons = screen.getAllByRole('backup-action-menu');
    expect(menuButtons[1]).toBeInTheDocument();
    await fireEvent.click(menuButtons[1]);

    // Take final table snapshot
    const finalTable = container.querySelector('.ant-table');
    expect(finalTable).toMatchSnapshot('final table state after deletion');
  }, 15000);

  it('should download a backup', async () => {
    jest.setTimeout(15000);

    const { container } = render(
      <Provider store={store}>
        <BackupPage />
      </Provider>
    );

    const initialTable = container.querySelector('.ant-table');
    expect(initialTable).toMatchSnapshot('initial table state');

    const menuButtons = screen.getAllByRole('backup-action-menu');
    expect(menuButtons[0]).toBeInTheDocument();
    await fireEvent.click(menuButtons[0]);

    // Take final table snapshot
    const finalTable = container.querySelector('.ant-table');
    expect(finalTable).toMatchSnapshot('final table state after download');
  }, 15000);
});
