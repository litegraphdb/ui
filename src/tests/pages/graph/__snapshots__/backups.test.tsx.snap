// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupPage should create a backup and should be visible in the table: add backup modal 1`] = `null`;

exports[`BackupPage should create a backup and should be visible in the table: backup name input 1`] = `
<input
  aria-required="true"
  class="ant-input css-dev-only-do-not-override-98ntnt ant-input-outlined"
  data-testid="filename-input"
  id="Filename"
  placeholder="Enter filename"
  type="text"
  value=""
/>
`;

exports[`BackupPage should create a backup and should be visible in the table: create backup button 1`] = `
<button
  class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
  style="font-weight: 500;"
  type="button"
>
  <span
    class="ant-btn-icon"
  >
    <span
      aria-label="plus-square"
      class="anticon anticon-plus-square"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="plus-square"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
        />
        <path
          d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
        />
      </svg>
    </span>
  </span>
  <span>
    Create Backup
  </span>
</button>
`;

exports[`BackupPage should create a backup and should be visible in the table: final table state 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should create a backup and should be visible in the table: initial table state 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should create a backup and should be visible in the table: submit button 1`] = `
<button
  class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-primary"
  disabled=""
  type="button"
>
  <span>
    OK
  </span>
</button>
`;

exports[`BackupPage should delete a backup and should be removed from the table: final table state after deletion 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger ant-dropdown-open"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should delete a backup and should be removed from the table: initial table state 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should display Create Backup button 1`] = `
<button
  class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-link"
  style="font-weight: 500;"
  type="button"
>
  <span
    class="ant-btn-icon"
  >
    <span
      aria-label="plus-square"
      class="anticon anticon-plus-square"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="plus-square"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
        />
        <path
          d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
        />
      </svg>
    </span>
  </span>
  <span>
    Create Backup
  </span>
</button>
`;

exports[`BackupPage should download a backup: final table state after download 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger ant-dropdown-open"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should download a backup: initial table state 1`] = `
<div
  class="ant-table css-dev-only-do-not-override-98ntnt"
>
  <div
    class="ant-table-container"
  >
    <div
      class="ant-table-content"
    >
      <table
        style="table-layout: auto;"
      >
        <colgroup>
          <col
            style="width: 100px;"
          />
        </colgroup>
        <thead
          class="ant-table-thead"
        >
          <tr>
            <th
              class="ant-table-cell react-resizable"
              scope="col"
            >
              Actions
              <span
                class="react-resizable-handle"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="ant-table-tbody"
        >
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="ant-table-row ant-table-row-level-0"
          >
            <td
              class="ant-table-cell"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-98ntnt ant-btn-text ant-btn-icon-only ant-dropdown-trigger"
                role="backup-action-menu"
                style="font-size: 16px;"
                type="button"
              >
                <span
                  class="ant-btn-icon"
                >
                  <span
                    aria-label="more"
                    class="anticon anticon-more"
                    role="img"
                    style="font-size: 20px;"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="more"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`BackupPage should render the BackupPage 1`] = `
<span
  class="ant-typography css-dev-only-do-not-override-98ntnt"
  data-testid="heading"
  style="font-weight: 600; font-size: 16px;"
>
  Backups
</span>
`;
